Specification draft

SamantazFox opened the pull request:
Very first draft of a specification

ChunkyProgrammer commented on README.md:
```diff
@@ -1,2 +1,105 @@
+Conditions for bumping major/minor/patch version: TBD
+Potential suffixes: `-draft`, TBD
+
+#### `"subscrpitions"`
```
```suggestion
#### `"subscriptions"`
```

ChunkyProgrammer commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+An array of Subscription objects. See below for details.
+
+#### `"subscrpitions"`
```
```suggestion
#### `"playlists"`
```

ChunkyProgrammer commented on README.md:
````diff
@@ -1,2 +1,105 @@
+```javascript
+{
+	"version": "X.Y.Z[-draft/-rc]",
+	"subscriptions": Subscrpitions[],
````
```suggestion
	"subscriptions": Subscriptions[],
```

ChunkyProgrammer commented on README.md:
```diff
@@ -1,2 +1,105 @@
+An array of `String`.
+
+Each item of the array is an URL representing one video in the playlist.
+Items has to be sorted in the same order as the videos in the playlist.
```
```suggestion
The array must be sorted in the same order as the videos in the playlist.
```

SamantazFox commented:
@ChunkyProgrammer Thanks ^^

FireMasterK commented on README.md:
````diff
@@ -1,2 +1,105 @@
+}
+```
+
+## Playlist object
````
A thumbnail attribute could be added too.

FireMasterK commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+#### > `"videos"`
+
+An array of `String`.
```
Could this instead be an object with and `id` and `url` for parsing simplicity?

SamantazFox replied:
I was thinking to evolve to [Uniform Resource Names (URN)](https://en.wikipedia.org/wiki/Uniform_Resource_Name), which can be easily adapted to various services and kinds of date. E.g:
* `urn:stdtube:youtube:channel:<id>`
* `urn:stdtube:youtube:video:<id>`
* `urn:stdtube:peertube:<instance FQDN>:<path>`

somerandomguy24 replied:
@SamantazFox  https://en.wikipedia.org/wiki/Uniform_Resource_Name#Namespaces
> In order to ensure the global uniqueness of URN namespaces, their identifiers (NIDs) are required to be registered with the [IANA](https://en.wikipedia.org/wiki/Internet_Assigned_Numbers_Authority).

Stypox replied:
We don't really need to pack information in only one string in a specific way. We can just define a json object of this form to replace URLs:
```json
{
  "type": "video",
  "service": "youtube",
  "instance": null,
  "id": "djV11Xbc914",
  "url": "https://www.youtube.com/watch?v=djV11Xbc914",
}
```

Also, if we decide not to use URLs, we should agree on what exactly an ID is for all services. For example, the first letters of YouTube playlist IDs are used to specify the playlist type, so technically those should not appear in `"id"`, but rather many different playlist `"type"`s should be created. Another example can be observed with PeerTube's accounts vs channels: they are (slightly) different entities with different URLs, so either we distinguish them into two different `"type"`s, or we include part of the URL in the ID. Examples: [YouTube normal playlist (PL)](https://www.youtube.com/playlist?list=RDCLAK5uy_nZiG9ehz_MQoWQxY5yElsLHCcG0tv9PRg), [YouTube genre mix (RDCLAK)](https://www.youtube.com/playlist?list=RDCLAK5uy_nZiG9ehz_MQoWQxY5yElsLHCcG0tv9PRg), [PeerTube account (/a/)](https://framatube.org/a/programmingfish@tube.piweb.be), [PeerTube channel (/c/)](https://framatube.org/c/programmingfish_channel@tube.piweb.be).

I think creating new `"type"`s each time a service is added is complicated, breaks compatibility and provides almost useless information. Therefore, only three types should exist: channel, playlist and video. The ID field should then contain standardized IDs based on the service (so e.g. PeerTube channel/account IDs would be, respectively, `c/...` and `a/...`).

... or we can just use only URLs directly and avoid all of these complications. In the end URLs are already "Uniform Resource Locator"s, so why bother creating our own way to uniquely locate resources? In this case we should also agree on which URL formats are ok, e.g. we might decide that these things are forbidden: useless parameters (e.g. t=5), short urls (e.g. youtu.be), http instead of https, YouTube urls from other frontends (e.g. https://invidious.kavin.rocks/watch?v=djV11Xbc914 instead of https://www.youtube.com/watch?v=djV11Xbc914)... Maintaining such a list of requirements is not so difficult, and even if one requirement were not respected by a frontend, other frontends would be able to open the URLs anyway in most cases, since usually frontends do allow opening strange/random URLs from the user anyway.

somerandomguy24 replied:
I personally prefer the json object approach rather than a url, as libraries for interacting with frontends/services typically do not have you pass a url, rather you may call high order functions with parameters to specify what data to return. if a url was used a (likely more complicated) parser would be required.

When it comes to the id, it may differ depending on what library is being used, so it is probably best to see what each of your libraries use.

FireMasterK replied:
Maybe we could include the `url` along with the `id`?  That seems like a more reasonable approach to me. I do agree that it can be tedious at times to parse just the id.

We can't just include just the `id`, as that contains no information about the service it is for.

somerandomguy24 replied:
In the example json object posted:
> ```
> {
>  "type": "video",
>  "service": "youtube",
>  "instance": null,
>  "id": "djV11Xbc914",
>  "url": "https://www.youtube.com/watch?v=djV11Xbc914",
>}
>```

The service, instance and type are specified.

FireMasterK commented on README.md:
````diff
@@ -1,2 +1,105 @@
+}
+```
+
+## Subscription object
````
Could we also include an `id` field too?

For YouTube, it would be the Channel ID.

SamantazFox replied:
See my previous comment regarding URNs

FireMasterK commented on README.md:
```diff
@@ -1,2 +1,105 @@
+The description of the playlist.
+Format: plain text (? TBD)
+
+CAN be ommited; In this case, the parser MUST assume an empty `String`.
```
nitpick: I feel null would be more appropriate if there's no description set/present

SamantazFox replied:
I personally prefer omitting fields than null-ing them

somerandomguy24 replied:
Json parsers in type/null-safe languages make it much easier/less likely to forget to check for null than an empty string.

Stypox replied:
I also agree it's better to use `null` instead of omitting the field completely

FireMasterK commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+The URL used to access the user subscribed channel.
+
+#### > `"thumbnail"`
+
+The URL to the user/channel's thumbnail.
```
I feel this should be an optional field, since:

- It's not really necessary for importing subscriptions.
- Can change and 404 in the future.

It's also ambiguous whether this should be a direct link to the image or a proxied URL to the media.

SamantazFox replied:
resolved in https://github.com/UniversalPipeWrench/unified-user-data-format/pull/2/commits/5a9c3d95466dab7a3b11d05be5159beb545a5621

FireMasterK reviewed (commented):
I also feel we should have a decoupled playlist and subscription files.

somerandomguy24 commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+#### > `"videos"`
+
+An array of `String`.
```
Probably not as these can just be fetched by the app, and just bloata the format for no good reason.

Stypox replied:
I think it should be consistent across the document, so either an (optional) thumbnail is allowed in all places, or it is not allowed anywhere

SamantazFox replied:
I've made the thumbnail present everywhere, but I don't think we should save more metadata in here.
I'm afraid that it'd grew out of control.

ghost commented:
> I also feel we should have a decoupled playlist and subscription files.

Or maybe just be able to choose which you want to export within the app.

somerandomguy24 commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+The visibility of the playlist.
+
+Accepted values: "public", "unlisted", "private"
```
There should also be a value for when the playlist is stored locally rather than on an external server.

SamantazFox replied:
This part of the spec only covers user-created playlists.
I think that saved/subscribed playlists (= not created by the user) should be stored differently.

SamantazFox replied:
Update: saved playlists have been addressed in https://github.com/UniversalPipeWrench/unified-user-data-format/pull/2/commits/d3f874b46ab88e8586b37a1c6ebedfb13c9d2e26

Stypox commented on README.md:
````diff
@@ -1,2 +1,105 @@
+}
+```
+
+## Playlist object
````
Something along these lines will do, I guess:

```json
{
  "name": "World Of Walker (Full Album)",
  "author": "Alan Walker",
  "url": "https://www.youtube.com/playlist?list=PLYT4vq6pQVSvgqA8Qsm8TXjYJMTT2XyNV",
  "thumbnail": "https://yt3.ggpht.com/ytc/abcdef..."
}
```

The author can be considered optional. Remote playlists should be in a separate section, obviously.

SamantazFox commented:
I've made the changes that were proposed in the various reviews above (may have missed some, sorry):
* Separated playlists into different files
* created a new object type (similar to subscriptions) for videos elements
* Added `type` (service name) and `id` (unique id) to subscription to ease parsing
* Added thumbnails to all objects, but made them optional
* Added a new file for "saved" (or "bookmarked") playlists

FireMasterK commented on README.md:
```diff
@@ -1,2 +1,106 @@
+Each string refers to the name of a playlist file (without the `playlist_` prefix)
+that was included in the export.
+
+#### `"watch_history_present"`
+
+A Boolean.
+
+Indicates whether the special playlist `watch_history` was included in the export.
```
Should we really be doing this? Watch history can have additional fields such as the watched till duration, and additional metadata. I would suggest an additional format to store this.

SamantazFox replied:
This field is only here to indicate if the `watch_history` file is part of the export.
The format of the watch history can still evolve regardless!

SamantazFox replied:
@FireMasterK if that confuses you, I can reword that part!
Right now, videos in history and regular playlists are the same objects, but that's definitely going to change ^^

FireMasterK replied:
> @FireMasterK if that confuses you, I can reword that part! Right now, videos in history and regular playlists are the same objects, but that's definitely going to change ^^

That makes more sense! I think they can be reworded once we finalize a specification for watch history too.

But, in such a scenario, can't we remove this field altogether?

SamantazFox replied:
I was thinking to make some files optional during the export (like the watch history).
This field's purpose is ti tell the importer whether that file is supposed to be present in the export.

somerandomguy24 commented on README.md:
```diff
@@ -1,2 +1,124 @@
+The description of the playlist.
+Format: plain text (? TBD)
+
+CAN be nil; In this case, the parser MUST assume an empty `String`.
```
Why is it assuming an empty string if the value is `nil`?

FireMasterK replied:
Agree, I think it's perfectly reasonable to assume `null` for a description if none was ever set.

SamantazFox replied:
I probably need to improve the wording... The idea is that nil and empty string have to be considered the same (that is, no description is present).

somerandomguy24 replied:
Ideally there should only be null if there is no description, and otherwise there is a string.

somerandomguy24 commented on spec/subscription_object.md:
There should be an `instance` field for federated services (peertube) which has a string representing the instance host (e.g. `peertube.example.org`). On a centralized service, it should be `null`/omitted (to my knowledge most json (de)serializers treat this as the same thing).

SamantazFox replied:
I don't think this is useful, as the `url` field already contain this information.
As far a my knowledge goes, most, if not all languages come with an URL parser, allowing to easily get the host part.
